### PR TITLE
Fix Sentry usage

### LINKED ISSUE #265: Evaluate options to replace Sentry

### LINKED ISSUE #305: Sentry is never initialized in the mediator
window.load is never raised here: https://github.com/mozilla/firefox-translations/blob/main/extension/mediator.js#L13

### LINKED ISSUE #309: Scrub level 3 data from sentry
Sentry collects page URL which is useful for error reporting but does not satisfy our current data collection policy. According to consent dialogue, we collect only technical and interaction data levels 1 and 2 by both telemetry and error reporting. URLs are level 3  “Stored Content & Communications” https://wiki.mozilla.org/Data_Collection.

So we should either scrub those from Sentry or change our data collection policy for error reporting. There are ways to scrub it client-side (preferred) or server side.

### PR DESCRIPTION
**Pull request by eu9ene**
- Remove unsafe dialog code from sentry
- Remove Sentry from content scripts and web worker, propagate exception to bgScript to report there instead

fixes #265 
fixes #305 
fixes #309 

I added test errors to different places in code and reporting works. I might have missed something, but the majority of the code is covered anyway.

I had to add a small lib for exception serialization to pass them in messages.

URLs are reported as smth like `[moz-extension://5239f6ed-8727-41c1-8aae-f03f037398b1/_generated_background_page.html]`, not actual page URLs.

### CODE REVIEW

**Comment by andrenatal** [comment]
@eu9ene Lgtm, did you test if the errors are still reported without these bits? 

Asking r? from @Rob--W too.

**Review by andrenatal — DISMISSED** [review]

**Comment by eu9ene** [comment]
> @eu9ene Lgtm, did you test if the errors are still reported without these bits?
> 
> Asking r? from @Rob--W too.

yes, I added some test errors and they are reported

**Comment by andrenatal** [comment]
> > @eu9ene Lgtm, did you test if the errors are still reported without these bits?
> > Asking r? from @Rob--W too.
> 
> yes, I added some test errors and they are reported

Great! So @Rob--W @wagnerand would this be enought?

**Comment by wagnerand** [comment]
cc @kewisch

**Comment by Rob--W** [comment]
I'm on PTO for the rest of this week, and this diff is too large to review from my phone, so I will defer to Andreas/Philipp for the completion of this review.

**Comment by wagnerand** [comment]
I have to say I am not comfortable approving this. We are monkey-patching a large machine generated file that is doing a lot (probably too many) things and is not really reviewable, in an extension that has privileged permissions.

**Comment by wagnerand** [comment]
> URLs are reported as smth like [moz-extension://5239f6ed-8727-41c1-8aae-f03f037398b1/_generated_background_page.html], not actual page URLs.

You still might want to want to strip the _host_ part of that url. It's an installation-unique identifier that we should not collect, especially not when using `web_accessible_resources`. Can you make that absolute URL a relative one, so you still have the path to the file the error occurred in, but don't collect `5239f6ed-8727-41c1-8aae-f03f037398b1`?

**Comment by eu9ene** [comment]
> > URLs are reported as smth like [moz-extension://5239f6ed-8727-41c1-8aae-f03f037398b1/_generated_background_page.html], not actual page URLs.
> 
> You still might want to want to strip the _host_ part of that url. It's an installation-unique identifier that we should not collect, especially not when using `web_accessible_resources`. Can you make that absolute URL a relative one, so you still have the path to the file the error occurred in, but don't collect `5239f6ed-8727-41c1-8aae-f03f037398b1`?

We can scrub it, not a problem. Sentry supports both server-side scrubbing with a regex or client side manually. So we can replace or remove `moz-extension://5239f6ed-8727-41c1-8aae-f03f037398b1` and leave only  `/_generated_background_page.html`. Would server-side scrubbing be sufficient?

**Comment by wagnerand** [comment]
> We can scrub it, not a problem. Sentry supports both server-side scrubbing with a regex or client side manually. So we can replace or remove moz-extension://5239f6ed-8727-41c1-8aae-f03f037398b1 and leave only /_generated_background_page.html. Would server-side scrubbing be sufficient?

If it doesn't cause a major issue, it might be better to filter it in the client.

**Comment by eu9ene** [comment]
> > We can scrub it, not a problem. Sentry supports both server-side scrubbing with a regex or client side manually. So we can replace or remove moz-extension://5239f6ed-8727-41c1-8aae-f03f037398b1 and leave only /_generated_background_page.html. Would server-side scrubbing be sufficient?
> 
> If it doesn't cause a major issue, it might be better to filter it in the client.

I added it on client side. It looks like this now:

```

url: /_generated_background_page.html

Error:

Test: TranslationWorker translate2

    /controller/translation/translationWorker.js in translate at line 591:7
    /controller/translation/translationWorker.js in consumeTranslationQueue/< at line 166:65

```

File: extension/model/static/errorReporting/sentry.js
Comment (kewisch) [inline extension/model/static/errorReporting/sentry.js:5725]: I can't get github to add a comment in context, so putting this here. Down in `UserAgent.prototype.setupOnce`, it seems an URL is collected. Can you check and scrub this URL as well?

File: extension/controller/backgroundScript.js
Comment (kewisch) [reply extension/controller/backgroundScript.js:37]: It looks like another way to do this would be to disable the history integration in the config, this would be preferred.

File: extension/controller/backgroundScript.js
Comment (eu9ene) [reply extension/controller/backgroundScript.js:35]: I will disable breadcrumbs integration for now. We can enable it back later after a more comprehensive audit.

**Comment by kewisch** [comment]
With above noted changes I would feel comfortable with using sentry.

File: extension/model/static/errorReporting/sentry.js
Comment (eu9ene) [reply extension/model/static/errorReporting/sentry.js:5725]: This method saves url to `event.request.url` and I already scrub it

File: extension/controller/backgroundScript.js
Comment (eu9ene) [reply extension/controller/backgroundScript.js:25]: I disabled breadcrumbs integration in the config, so there is no place where history artifacts can be present now:
It's what it sends:
```
{
  "exception": {
    "values": [
      {
        "type": "TypeError",
        "value": "messages.test is not a function",
        "stacktrace": {
          "frames": [
            {
              "filename": "/controller/translation/translationWorker.js",
              "function": "consumeTranslationQueue/<",
              "in_app": true,
              "lineno": 166,
              "colno": 65
            },
            {
              "filename": "/controller/translation/translationWorker.js",
              "function": "translate",
              "in_app": true,
              "lineno": 591,
              "colno": 30
            }
          ]
        },
        "mechanism": {
          "type": "generic",
          "handled": true
        }
      }
    ]
  },
  "level": "error",
  "event_id": "ea15cd77807d4e05ba49d0ebfab6d06e",
  "platform": "javascript",
  "timestamp": 1651702973.263,
  "environment": "production",
  "release": "firefox-translations@1.0",
  "sdk": {
    "integrations": [
      "InboundFilters",
      "FunctionToString",
      "TryCatch",
      "GlobalHandlers",
      "LinkedErrors",
      "Dedupe",
      "UserAgent"
    ],
    "name": "sentry.javascript.browser",
    "version": "6.19.2",
    "packages": [
      {
        "name": "npm:@sentry/browser",
        "version": "6.19.2"
      }
    ]
  },
  "request": {
    "url": "/_generated_background_page.html",
    "headers": {
      "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:102.0) Gecko/20100101 Firefox/102.0"
    }
  },
  "tags": {
    "skippedNormalization": true
  },
  "extra": {
    "normalizeDepth": 3
  }
}
```

File: extension/controller/backgroundScript.js
Comment (eu9ene) [reply extension/controller/backgroundScript.js]: added http/https, but I do not observe any actual URLs there except moz-extension:// ones

File: extension/controller/backgroundScript.js
Comment (eu9ene) [reply extension/controller/backgroundScript.js]: done

**Comment by kewisch** [comment]
Ah one last comment (sorry, it is late!). sentry makes a lot of changes to monkeypatch event listeners and other things. If you are using this in a privileged context, you need to make absolutely sure that sentry will not load in a Firefox browser chrome window. Sentry use should be contained to separate DOM documents/windows only used for Firefox Translations.

File: extension/controller/backgroundScript.js
Comment (eu9ene) [reply extension/controller/backgroundScript.js:13]: The only scenario when an unhandled error is not passed through my scrubber is when the error happens in the scrubber itself. I added returning null to not report anything in this case.

File: extension/controller/backgroundScript.js
Comment (eu9ene) [reply extension/controller/backgroundScript.js]: also added deleting `event.request.url` completely, it wasn't very useful anyway

File: extension/model/static/errorReporting/sentry.js
Comment (eu9ene) [reply extension/model/static/errorReporting/sentry.js:5725]: I decided to delete `event.request.url` completely, it wasn't very useful anyway

**Comment by andrenatal** [comment]
> Ah one last comment (sorry, it is late!). sentry makes a lot of changes to monkeypatch event listeners and other things. If you are using this in a privileged context, you need to make absolutely sure that sentry will not load in a Firefox browser chrome window. Sentry use should be contained to separate DOM documents/windows only used for Firefox Translations.

@kewisch Sentry will exit only in the background script pertinent to the extension, and the error reports will be sent to it via messaging, so it won't be loaded anywhere else. I assume that satisfies your requirement ?

**Comment by eu9ene** [comment]
> Ah one last comment (sorry, it is late!). sentry makes a lot of changes to monkeypatch event listeners and other things. If you are using this in a privileged context, you need to make absolutely sure that sentry will not load in a Firefox browser chrome window. Sentry use should be contained to separate DOM documents/windows only used for Firefox Translations.

We load it only in the background script now. We do not use it in experiments API. As far as I understand it is the only way to interfere with Firefox chrome window.

**Review by andrenatal — APPROVED** [review]

**Comment by andrenatal** [comment]
We need to merge this to unblock the other issue, but @kewisch please let us know if there's something here you want to revisit.

File: extension/controller/backgroundScript.js
Comment (wagnerand) [reply extension/controller/backgroundScript.js]: Nit: Good practice would be to have an early

```js
if (!ev.request.url.startsWith("moz-extension://") {
// Either don't send this sentry report at all since the error is not part of the extension
// or at least do
ev.request.url = null;
}
```

**Pull request merged**
